packages/ozone/src/set/service.ts

```diff
@@ -0,0 +1,203 @@
+    sortBy: 'name' | 'createdAt' | 'updatedAt'
+  }): Promise<{
+    sets: Selectable<OzoneSet & { setSize: number }>[]
+    cursor?: string
+  }> {
+    let qb = this.db.db
+      .selectFrom('ozone_set as s')
+      .leftJoin(
+        (eb) =>
+          eb
+            .selectFrom('ozone_set_value')
+            .select(['setId'])
+            .select((eb) => eb.fn.count('id').as('count'))
+            .groupBy('setId')
+            .as('sv'),
+        (join) => join.onRef('sv.setId', '=', 's.id'),
+      )
+      .select([
+        's.id',
+        's.name',
+        's.description',
+        's.createdAt',
+        's.updatedAt',
+        sql<number>`coalesce(sv.count, 0)`.as('setSize'),
+      ])
+      .limit(limit)
+
+    if (namePrefix) {
+      qb = qb.where('s.name', 'like', `${namePrefix}%`)
+    }
+
+    if (cursor) {
+      if (sortBy === 'name') {
+        qb = qb.where('s.name', '>', cursor)
+      } else {
+        qb = qb.where(`s.${sortBy}`, '>', new Date(cursor))
+      }
+    }
+
+    qb = qb.orderBy(`s.${sortBy}`, 'asc')
+
+    const sets = await qb.execute()
+    const lastItem = sets.at(-1)
+
+    return {
+      sets,
+      cursor: lastItem
+        ? sortBy === 'name'
+          ? lastItem?.name
+          : lastItem?.[sortBy].toISOString()
+        : undefined,
+    }
+  }
+
+  async getByName(
+    name: string,
+    txn: Database,
+  ): Promise<Selectable<OzoneSet> | undefined> {
+    const query = txn.db
+      .selectFrom('ozone_set')
+      .selectAll()
+      .where('name', '=', name)
+
+    return await query.executeTakeFirst()
+  }
+
+  async getSetWithValues({
+    name,
+    limit,
+    cursor,
+  }: {
+    name: string
+    limit: number
+    cursor?: string
+  }): Promise<
+    { set: Selectable<OzoneSet>; values: string[]; cursor?: string } | undefined
+  > {
+    const set = await this.getByName(name, this.db)
+    if (!set) return undefined
+
+    let qb = this.db.db
+      .selectFrom('ozone_set_value')
+      .select(['value'])
+      .where('setId', '=', set.id)
+      .limit(limit)
+
+    if (cursor) {
+      qb = qb.where('value', '>', cursor)
+    }
+
+    qb = qb.orderBy('value', 'asc')
+
+    const values = await qb.execute()
+
+    return {
+      set,
+      values: values.map((v) => v.value),
+      cursor: values.at(-1)?.value,
+    }
+  }
+  async upsert(
+    { name, description }: Pick<OzoneSet, 'name' | 'description'>,
+    txn: Database,
+  ): Promise<Selectable<OzoneSet>> {
+    const query = txn.db
+      .insertInto('ozone_set')
+      .values({
+        name,
+        description,
+        updatedAt: new Date(),
+      })
+      .onConflict((oc) =>
+        oc.column('name').doUpdateSet({
+          description,
+          updatedAt: new Date(),
+        }),
+      )
+      .returningAll()
+
+    return await query.executeTakeFirstOrThrow()
+  }
+
+  async addValues(
+    setId: number,
+    values: string[],
+    txn: Database,
+  ): Promise<void> {
+    const query = txn.db
```

Comment:
If `txn` this must be a transaction for correctness / data integrity it can be enforced here:
```suggestion
    txn: Database,
  ): Promise<void> {
    txn.assertTransaction()
    const query = txn.db
```

Reply:
ended up moving the transactions inside the methods since we moved all the additional checks out of the transactions.